### PR TITLE
Project 03: Implement Model Validations and Specs

### PR DESCRIPTION
**Pull request by GhausAliKhan**
### In this branch I did:

The changes ensure that our models enforce data integrity and work as expected.

- Validations added to the `User` model to ensure that the name is present and the posts counter is a non-negative integer.
- Validations added to the `Post` model to ensure that the title is present, within 250 characters, and that both comments and likes counters are non-negative integers.
- Comprehensive unit tests for the methods and validations in the `User`, `Post`, `Like`, and `Comment` models.

### New Files & Modifications
- The User and Post models now have validations that align with the project's requirements.
- Unit tests have been written using RSpec to verify that all model methods perform correctly and handle edge cases as expected.

### Conclusion
This ensures that the Blog App's models are robust, with validations preventing invalid data entry and unit tests ensuring expected behavior.

### CODE REVIEW

File: spec/models/post_spec.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+      it 'returns the 5 most recent comments' do
+        recent_comments = @post.recent_comments
+        expect(recent_comments.count).to eq(5)
+        expected_comments = @post.comments.order(created_at: :desc).limit(5)
+        expect(recent_comments).to match_array(expected_comments)
+      end
+    end
+  end
+end
```
Comment (ghost) [inline spec/models/post_spec.rb]: - [x] 🌟 Awesome job on summarizing your tests! Could we also test the `increment_user_posts_counter` method in your `Post model`? Thanks a bunch!

#### Keep up the good work 👏

**Review by frank1738 — CHANGES_REQUESTED** [review]
Hi @GhausAliKhan , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!




## Highlights

✅ Most of the Models are tested correctly
✅ Effectively Applied rails callback methods👍 



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

N/A
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/post_spec.rb
Comment (shella12) [inline spec/models/post_spec.rb:1]: - [ ] According to the [project requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.) any custom methods in the models should have a corresponding unit test. Please make sure to include a unit test for `increment_user_posts_counter` method in your `Post` model as requested by the previous reviewer.

Comment (shella12) [inline spec/models/post_spec.rb:1]: - [x] According to the [project requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_validations_model_spec.md#:~:text=Add%20unit%20specs%20for%20all%20of%20your%20models%27%20methods%20and%20validations.) any custom methods in the models should have a corresponding unit test. Please make sure to include a unit test for `increment_user_posts_counter` method in your `Post` model as requested by the previous reviewer.

File: spec/models/post_spec.rb (hunk 1 above)
Comment (GhausAliKhan) [reply spec/models/post_spec.rb]: Sure I'll add it. Thanks for the review

**Review by shella12 — CHANGES_REQUESTED** [review]
Hi @GhausAliKhan, 

Great work on making the changes requested by a previous reviewer 👏 👏 👏 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet. You are almost there...



## Highlights

- All linters passed ✅ 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/post_spec.rb
Comment (GhausAliKhan) [reply spec/models/post_spec.rb:1]: @shella12 
I already added the Unit test, can you check again?

Comment (shella12) [reply spec/models/post_spec.rb:1]: Hi @GhausAliKhan By the time I reviewed your PR you had not made any changes requested by the previous reviewer so I had to request the changes again and had already submitted the review.  While giving the feedback we can't really check for any new commits. Its highly recommended not to make any changes after submission of PR so that the review is consistent and relevant to the submitted PR. 

Since the review has already been submitted you would have to resubmit the PR through dashboard.

**Review by gemmen29 — APPROVED** [review]
Hi @GhausAliKhan, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
